<a href="https://colab.research.google.com/github/rahiakela/machine-learning-research-and-practice/blob/main/grokking-machine-learning/08-naive-bayes-model/01_spam_detection_with_naive_bayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Spam-detection with naive Bayes

Now that we have developed the algorithm, let’s roll up our sleeves and code the naive Bayes
algorithm.

##Setup

In [1]:
import pandas as pd
import numpy as np
import random
import matplotlib.pyplot as plt

random.seed(0)

In [ ]:
!wget https://github.com/luisguiserrano/manning/raw/master/Chapter_8_Naive_Bayes/emails.csv

##Dataset preprocessing

First, let's load the dataset.

In [3]:
emails = pd.read_csv("emails.csv")
emails.head()

text  spam
0  Subject: naturally irresistible your corporate...     1
1  Subject: the stock trading gunslinger  fanny i...     1
2  Subject: unbelievable new homes made easy  im ...     1
3  Subject: 4 color printing special  request add...     1
4  Subject: do not have money , get software cds ...     1

In [4]:
def process_email(text):
  text = text.lower()
  return list(set(text.split()))

In [5]:
emails["words"] = emails["text"].apply(process_email)
emails.head()

text  spam  \
0  Subject: naturally irresistible your corporate...     1   
1  Subject: the stock trading gunslinger  fanny i...     1   
2  Subject: unbelievable new homes made easy  im ...     1   
3  Subject: 4 color printing special  request add...     1   
4  Subject: do not have money , get software cds ...     1   

                                               words  
0  [market, do, %, original, interested, full, ma...  
1  [inflexible, nameable, morristown, and, try, o...  
2  [all, factor, form, and, take, website, ., is,...  
3  [golden, canyon, version, :, ramsey, form, mai...  
4  [me, do, all, finish, ain, comedies, are, ., i...

##Finding the priors

Let’s first find the probability that an email is spam (the prior).

In [6]:
num_emails  = len(emails)
num_spam = sum(emails["spam"])
print(f"Number of emails: {num_emails}")
print(f"Number of spam emails: {num_spam}")
print()

# Calculating the prior probability that an email is spam
print(f"Probability of spam: {num_spam / num_emails}")

Number of emails: 5728
Number of spam emails: 1368

Probability of spam: 0.2388268156424581


We deduce that the prior probability that the email is spam is around 0.24. This is the probability
that an email is spam if we don’t know anything about the email. 

Likewise, the prior probability
that an email is ham is around 0.76.

##Finding the posteriors